### PR TITLE
Fix Robertson Calibration NaN Bug

### PR DESCRIPTION
**Pull request by danielenricocahall**
**Merge with extra**: https://github.com/opencv/opencv_extra/pull/873

Addresses https://github.com/opencv/opencv/issues/18180. The issue seemed to be with numerical stability when `w_sum` matrix contained zeroes. Should be merged in conjunction with https://github.com/opencv/opencv_extra/pull/873, which contains test data taken from original issue.
### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [x] The PR is proposed to proper branch
- [x] There is reference to original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

File: modules/photo/test/test_hdr.cpp [hunk 1]
```diff
@@ -249,4 +249,22 @@ TEST(Photo_CalibrateRobertson, regression)
...
+    for (auto & i : fn)
+        images.push_back(imread(i));
+    vector<float> times {15, 2.5, 0.25, 0.333333};
+    Mat response, expected;
+    Ptr<CalibrateRobertson> calibrate = createCalibrateRobertson(2, 0.01);
+    calibrate->process(images, response, times);
+    Mat response_no_nans = response.clone();
+    patchNaNs(response_no_nans);
+    // since there should be no NaNs, original response vs. response with NaNs patched should be identical
+    bool isEqual = (sum(response != response_no_nans) == Scalar(0,0,0));
```
Comment (alalek) [inline modules/photo/test/test_hdr.cpp:266]: `norm(response, response_no_nans, NORM_INF) == 0.0` ?

File: modules/photo/test/test_hdr.cpp [hunk 2]
```diff
@@ -249,4 +249,22 @@ TEST(Photo_CalibrateRobertson, regression)
     checkEqual(expected, response, 1e-1f, "CalibrateRobertson");
 }
 
+TEST(Photo_CalibrateRobertson, bug_18180)
+{
+    vector<Mat> images;
+    vector<cv::String> fn;
+    glob(string(cvtest::TS::ptr()->get_data_path()) + "hdr/exposures/bug_18180/*.jpg", fn, false);
+    for (auto & i : fn)
+        images.push_back(imread(i));
```
Comment (alalek) [inline modules/photo/test/test_hdr.cpp:258]: Please avoid using `glob()` in tests. Use fixed input instead.

File: modules/photo/test/test_hdr.cpp (hunk 1 above)
Comment (danielenricocahall) [reply modules/photo/test/test_hdr.cpp:266]: That's smarter - let me do that. Thanks!

File: modules/photo/test/test_hdr.cpp (hunk 2 above)
Comment (danielenricocahall) [reply modules/photo/test/test_hdr.cpp:258]: Got it!

**Comment by danielenricocahall** [comment]
> @danielenricocahall Please take a look on failed builds (you may start from "debug" local build)

I think all issues are resolved - please feel free to re-review at earliest convenience.

File: modules/photo/test/test_hdr.cpp (hunk 1 above)
Comment (alalek) [reply modules/photo/test/test_hdr.cpp]: NORM_INF doesn't properly handle NaNs.

There is still a problem:

```
EXPECT_EQ(0.0, cv::norm(response, response_no_nans, NORM_L2));
```

Results to this error message (Linux Debug):

```
Expected equality of these values:
  0.0
    Which is: 0
  cv::norm(response, response_no_nans, NORM_L2)
    Which is: -nan
[  FAILED  ] Photo_CalibrateRobertson.bug_18180 (1789 ms)
```

Comment (danielenricocahall) [reply modules/photo/test/test_hdr.cpp]: Ah I figured out the issue, I had to move the epsilon back. Thank you!

**Comment by vpisarev** [comment]
thank you! :+1:

**Review by vpisarev — APPROVED** [review]

File: modules/photo/src/merge.cpp [hunk 3]
```diff
@@ -342,7 +342,7 @@ class MergeRobertsonImpl CV_FINAL : public MergeRobertson
             LUT(images[i], response, im);
 
             result += times.at<float>((int)i) * w.mul(im);
-            wsum += times.at<float>((int)i) * times.at<float>((int)i) * w;
+            wsum += times.at<float>((int)i) * times.at<float>((int)i) * (w + DBL_EPSILON);
         }
         result = result.mul(1 / wsum);
```
Comment (alalek) [inline modules/photo/src/merge.cpp:347]: Patching `wsum` doesn't look valid.

Patch divide operation instead (`Scalar::all()` is necessary as we work with multi-channel data):
```
result = result.mul(1 / (wsum + Scalar::all(DBL_EPSILON)));
```

Comment (danielenricocahall) [reply modules/photo/src/merge.cpp]: @alalek Ah good catch. It technically works, but a warning pops up:

```
assign OpenCV/MatExpr: processing of multi-channel arrays might be changed in the future: https://github.com/opencv/opencv/issues/16739
```

Addressed the issue and actually moved the epsilon to where the division occurs. Thank you!

**Pull request merged**